src/goto-programs/lazy_goto_model.h

```diff
@@ -51,8 +74,86 @@ class lazy_goto_modelt
   /// \returns The frozen goto_modelt or an empty optional if freezing fails
   static std::unique_ptr<goto_modelt> freeze(lazy_goto_modelt &&model)
   {
+    if(!model.freeze())
+      return nullptr;
     return std::move(model.goto_model);
   }
+
+private:
+  bool freeze();
 };
 
+
+template<typename THandler>
+lazy_goto_modelt lazy_goto_modelt::from_handler_object(
+  THandler &handler,
+  const optionst &options,
+  message_handlert &message_handler)
+{
+  return lazy_goto_modelt(
+    [&handler, &message_handler] (
+      const irep_idt &function_name,
+      goto_functionst::goto_functiont &function,
+      symbol_tablet &symbol_table) -> void
+    {
+      messaget msg(message_handler);
+      try
+      {
+        handler.process_goto_function(function_name, function, symbol_table);
+      }
+      catch(const char *e)
+      {
+        msg.error() << "process_goto_function: " << e << messaget::eom;
+        throw;
+      }
+      catch(const std::string &e)
+      {
+        msg.error() << "process_goto_function: " << e << messaget::eom;
+        throw;
+      }
+      catch(int e)
+      {
+        msg.error()
+          << "process_goto_function: Numeric exception : " << e
```

Comment:
Side comment: this makes little sense, but I defer the refactoring of this to https://github.com/diffblue/cbmc/issues/1314.